feat(cli): pass CloudFormation parameters to "cdk deploy"

dkanhakl opened the pull request:
This commit closes #1237

I have added the command line option --parameters, -P which allows for
adding CloudFormation parameters at deploy time.

The functionality for deploying with parameters was already present,
but I have just added a way for the command line to pass it to the
deployment code.

The motivation for adding this functionality has been that we have
a setup where we share an account with access to CloudFormation, which
is why we do not want to have some fields in plain text, while they do
not have access to other services, which did not support secrets within
the fields we want to keep secret. This allows us to have parameters
that can be kept secret from the templates, while also having parameters
with NoEcho.

I have verified this functionality by both deploying a cdk project
with a CfnParameter from @aws-cdk/core, and an original CloudFormation
template with parameters.



----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3223926337dfd3a4e95cf8030a6aa9928a7a6151
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0d471d5b-7474-4671-9f09-73b3c9300571%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b19567457d51172055ec8df16a9324de5fd615c4
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=46df1fb5-f81d-4bab-8381-4210a48b2b41%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/aws-cdk/bin/cdk.ts:
```diff
@@ -69,6 +69,7 @@ async function parseCommandLineArguments() {
       .option('tags', { type: 'array', alias: 't', desc: 'Tags to add to the stack (KEY=VALUE)', nargs: 1, requiresArg: true })
       .option('execute', { type: 'boolean', desc: 'Whether to execute ChangeSet (--no-execute will NOT execute the ChangeSet)', default: true })
       .option('force', { alias: 'f', type: 'boolean', desc: 'Always deploy stack even if templates are identical', default: false })
+      .option('parameters', { type: 'array', alias: 'P', desc: 'Additional parameters passed to CloudFormation at deploy time (KEY=VALUE)', nargs: 1, requiresArg: true, default: {} })
```
all our other options are aliased to lower case. Any reason this needs to be `P`?

dkanhakl replied:
the lower `p` is used for plugins, which makes this crash. Unfortunately. :)

shivlaks replied:
I think it's better to leave aliases out for the time being and we can introduce it later if the need really arises in that case. thoughts?

dkanhakl replied:
That could work. I only added it to fit in with the other options.

dkanhakl commented:
- Currently looking at the `cdk diff`, it does not appear to account for parameters.

- have not tried nested stacks yet, as I do not have a working example. (Still pretty new to cdk)

- When looking at the README for the cli module I did not find any section where other cli parameters are mentioned, which is why I have omitted it for now. But if you know where the documentation would be fitting, I would love to write some.

- I did not know you could do wild card deploy, but after testing it, it looks like this implementation does not work, as it tries to duplicate the parameters. Do you have any preference to the syntax for choosing the stack that a parameter belongs to? My suggestion would be --parameters "StackName:Key=Value"

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d6c89af7417bc981f69d3470a61a983ce5c97fa7
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5a012c21-e6bf-4537-970b-95f4989b93f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 88658d712eb4b046a7f9a64ca7814af87640d9e0
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=351becc0-5584-4b83-b28a-106cf2d03de8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dkanhakl commented:
I have added some tests and implemented the syntax for specifying which stack a parameter belongs to.
Please let me know what would be the next step for this feature, as I will continue working on this tomorrow.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: dc5a0f372ff23ffea24037c06cde15279e8aba54
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c61a3c91-c30d-4f50-9d8e-d7ed4a9c7199%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

john-tipper commented:
What’s needed to get this over the line please? It looks like we’re pretty close, no? We’re extremely keen to see this merged as it’ll open up using CDK for us as this particular issue is blocking us. Please let me know if I can help.

dkanhakl commented:
Thank you for your interest @john-tipper currently we need to check whether this works with nested stacks, and we need to consider if this should impact `cdk diff` or not. If it should, then that needs some work.
I am currently looking into making it work with diff, as when you try to deploy an identical stack with new parameters it still shows up as not changed, and needs a --force in order to be redeployed.

shivlaks commented:
> Thank you for your interest @john-tipper currently we need to check whether this works with nested stacks, and we need to consider if this should impact `cdk diff` or not. If it should, then that needs some work.
> I am currently looking into making it work with diff, as when you try to deploy an identical stack with new parameters it still shows up as not changed, and needs a --force in order to be redeployed.

* I think it's okay to park `cdk diff` support for now unless we feel there's going to be unintended consequences from using parameters. 
* For wild card support, I think your suggestion is reasonable... might also be simpler to only support parameters when a stack is specified and not with wild-card deploys (i.e. add the capability to specify stack as an incremental feature)
* Did you get a chance to see how this behaves with Nested Stacks yet?

dkanhakl commented:
I have implemented the version of the parameter parsing that I mentioned above, so if that is fine, then I think we could keep that.
I will now switch from working on the diff "issue", as this seems to be a bit more work, to testing the nested stack thing. Also just add a few lines to the readme.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: df634897d5476aa4543b3f10466a791e8f53333e
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b7fe4630-6268-49f0-a765-017270aa5bd0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 237e296ee1e3c7bc617a9af3318426455d0ea963
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d61b94e1-d009-4d2d-b845-cf6559b589ad%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3ea24d6f44d9c6ad2c026fccea12b0348ad78402
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=055caf8d-2a2e-4b0f-bc8e-1db889b02df0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented:
@john-tipper i'll also be hopping on resolving the remaining threads on this issue. i'd love to get it into the next release so i'm trying to make that happen

dkanhakl commented:
@shivlaks is there anything I can do to help make this happen?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e54ba0551a3369e1ad082bba15328e6fef94cb3e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e22773c8-b96d-47d4-b0bc-601b241971b9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 73b783b759053284892d01033f05478053e41d5c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2afad382-7b56-4d8c-a1dc-ed06824bf01f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented:
update: I spent some time testing it out

* I think the convention of prefixing with `--parameters "StackName:Key=Value"` seems reasonable to me
* Nested stacks are fine and can consume parameters through the parent stack
* `cdk diff` doesn't account for parameters - which may be reasonable for now

to resolve:
* how do we *remove* parameter from a stack?
* since `diff` won't show parameters, will a change in parameters clobber previously existing parameters without warning? (should it?)

thoughts @john-tipper  @dkanhakl

joehillen commented on packages/aws-cdk/README.md:
```diff
@@ -133,6 +133,12 @@ currently deployed stack to the template and tags that are about to be deployed
+In order to pass parameters to your template during deployment, you can use `--parameters
+(STACK:KEY=VALUE)`. This will apply the value `VALUE` to the key `KEY` for stack `STACK`. The stack
+name can either be `*` or omitted in order to apply to all stacks.
+Note: The parameters will not propagate to NestedStacks. These has to be sent with the constructor
```
s/These has to/These have to/

or "These must be" would be better

shivlaks replied:
agreed, an example with a small snippet of usage would be nice too.

joehillen commented:
> will a change in parameters clobber previously existing parameters without warning? (should it?)

I was just about to ask if this would happen. I don't think it should. I'm dealing with this right now with a stack that has a default parameter that is updated by an external event. I only want the default used for the initial stack creation. Whenever the stack template updates, I want it to use the existing non-default value.

shivlaks commented:
> > will a change in parameters clobber previously existing parameters without warning? (should it?)
> 
> I was just about to ask if this would happen. I don't think it should. I'm dealing with this right now with a stack that has a default parameter that is updated by an external event. I only want the default used for the initial stack creation. Whenever the stack template updates, I want it to use the existing non-default value.

I agree that anything which is surprising to a customer without warning is not ideal (unless they're using the `--force` flag. I'm going to give this a quick test in a couple of hours and see what we can do about that. I'd love to resolve this one, but I see this is the primary remaining thing for getting this first commit in. I do think there should be a mechanism to remove a parameter as well (maybe it's as simple as proving an empty value to an existing key?)

dkanhakl commented:
Thank you for the feedback and interest. I need to understand what you mean by removing a parameter. 

First of all, if I understand it correctly, regarding default parameters. If you have a default and apply a parameter from the cli, then it should follow the behaviour of cloudformation and override the default. (which is what should be implemented already)

But currently if you do not change anything in the template, then it will not detect a redeploy with different parameters as a change, and does need a `--force` parameter to actually install the new parameters.

dkanhakl commented:
> > will a change in parameters clobber previously existing parameters without warning? (should it?)
> 
> I was just about to ask if this would happen. I don't think it should. I'm dealing with this right now with a stack that has a default parameter that is updated by an external event. I only want the default used for the initial stack creation. Whenever the stack template updates, I want it to use the existing non-default value.

I have not tested the scenario that you mention here. Since I am still new to AWS, could you please elaborate how cloudformation handles this, or point me to some documentation of this behaviour?

The current implementation has only added that the ability to pass parameters to the deployment command, and not changed anything to the actual deployment itself. Does anyone have any knowledge of how this has been handled with the parameters that cdk already uses?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2024a54f80bdc0759b117941698ec13059ab6e36
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=50d79216-4ebb-43c9-ae31-ace59fb76eaf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 98b655ad6451a65dc7b256bc39cb626392022dfb
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3e8009e5-c62e-4530-bd7b-af8ac1b245fb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 65e3cc21d9ca4b95bca81ff20dfbc2516bab0e6e
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=af0c49f9-c243-4d1b-b6a7-0b8e6645fef8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

joehillen commented:
> I have not tested the scenario that you mention here. Since I am still new to AWS, could you please elaborate how cloudformation handles this, or point me to some documentation of this behaviour?
> 
> The current implementation has only added that the ability to pass parameters to the deployment command, and not changed anything to the actual deployment itself. Does anyone have any knowledge of how this has been handled with the cloudformation parameters that cdk already uses?

It seems to be expected behavior for CF even though it's surprising and frustrating (typical for CF). If update-stack is sent to CF without parameters, then the default is used even if the parameters were set in a previous `update-stack` call. I think the fix would be to lookup parameters that were set in the stack and then pass them again with the next stack update.

It's probably outside the scope of this PR. I'll open another PR after this one is merged.

joehillen commented:
For some more context, this has been a long standing feature request for the AWS CLI https://github.com/aws/aws-cli/issues/2589

shivlaks commented:
> It seems to be expected behavior for CF even though it's surprising and frustrating (typical for CF). If update-stack is sent to CF without parameters, then the default is used even if the parameters were set in a previous `update-stack` call. I think the fix would be to lookup parameters that were set in the stack and then pass them again with the next stack update.
> 
> It's probably outside the scope of this PR. I'll open another PR after this one is merged.

@joehillen do you mind creating a feature-request issue for that functionality? I agree that it should be addressed in a different PR

@dkanhakl this is looking good, I'm running some final tests and will make some quick tweaks to the README, but otherwise I think we're looking good to move this one forward! .

shivlaks commented:
> > will a change in parameters clobber previously existing parameters without warning? (should it?)
> 
> I was just about to ask if this would happen. I don't think it should. I'm dealing with this right now with a stack that has a default parameter that is updated by an external event. I only want the default used for the initial stack creation. Whenever the stack template updates, I want it to use the existing non-default value.

It doesn't currently provide a helpful error message indicating that you've tried to do this. I'd like to address that in a follow-on PR as I think it would be helpful to detect and provide a helpful message indicating that it can be replaced by using `--force`

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 775b1aa1e40ba038a6cac515b8ef835c656bc1e8
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c5648496-c143-4932-9651-22156008778e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 82ce5cfb71f49d75c68fb3d43dd36480930bac53
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=76d7f867-46d7-4e4b-96c2-2fdd63abe6ab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 521c3137cd3faffcbf1119316c22cb80ab17c0d3
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=579ea82d-806e-434f-aad6-0ca92116c347%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
